test/screenshot/lib/image-differ.js

```diff
@@ -0,0 +1,167 @@
+      if (!expectedCapture) {
+        continue;
+      }
+
+      pagePromises.push(
+        this.compareOnePage_({
+          htmlFilePath,
+          actualCapture,
+          expectedCapture,
+        })
+      );
+    }
+
+    // Flatten the array of arrays
+    const diffResults = [].concat(...(await Promise.all(pagePromises)));
+
+    // Filter out images with no diffs
+    return diffResults.filter((diffResult) => Boolean(diffResult.diffImageBuffer));
+  }
+
+  /**
+   * @param {string} htmlFilePath
+   * @param {!CaptureJson} actualCapture
+   * @param {!CaptureJson} expectedCapture
+   * @return {!Promise<!Array<!ImageDiffJson>>}
+   * @private
+   */
+  async compareOnePage_({
+    htmlFilePath,
+    actualCapture,
+    expectedCapture,
+  }) {
+    /** @type {!Array<!Promise<!ImageDiffJson>>} */
+    const imagePromises = [];
+
+    const actualScreenshots = actualCapture.screenshots;
+    const expectedScreenshots = expectedCapture.screenshots;
+
+    for (const [browserKey, actualImageUrl] of Object.entries(actualScreenshots)) {
+      // Screenshot image for this browser is not present in `golden.json` on `master`
+      const expectedImageUrl = expectedScreenshots[browserKey];
+      if (!expectedImageUrl) {
+        continue;
+      }
+
+      imagePromises.push(
+        this.compareOneImage_({actualImageUrl, expectedImageUrl})
+          .then(
+            (diffImageBuffer) => ({
+              htmlFilePath,
+              browserKey,
+              expectedImageUrl,
+              actualImageUrl,
+              diffImageUrl: null, // populated by `Controller`
+              diffImageBuffer,
+            }),
+            (err) => Promise.reject(err)
+          )
+      );
+    }
+
+    return Promise.all(imagePromises);
+  }
+
+  /**
+   * @param {string} actualImageUrl
+   * @param {string} expectedImageUrl
+   * @return {!Promise<?Buffer>}
+   * @private
+   */
+  async compareOneImage_({
+    actualImageUrl,
+    expectedImageUrl,
+  }) {
+    console.log(`➡ Comparing snapshot to golden: "${actualImageUrl}" vs. "${expectedImageUrl}"...`);
+
+    const [actualImageBuffer, expectedImageBuffer] = await Promise.all([
+      this.imageCache_.getImageBuffer(actualImageUrl),
+      this.imageCache_.getImageBuffer(expectedImageUrl),
+    ]);
+
+    const diffResult = await this.computeDiff_({
+      actualImageBuffer,
+      expectedImageBuffer,
+    });
+
+    if (diffResult.rawMisMatchPercentage < 0.01) {
+      console.log(`✔ No diffs found for "${actualImageUrl}"!`);
+      return null;
+    }
+
+    console.log(`✗︎ Image "${actualImageUrl}" has changed!`);
+    return diffResult.getBuffer();
+  }
+
+  /**
+   * @param {!Buffer} actualImageBuffer
+   * @param {!Buffer} expectedImageBuffer
+   * @return {!Promise<!ResembleApiComparisonResult>}
+   * @private
+   */
+  async computeDiff_({
+    actualImageBuffer,
+    expectedImageBuffer,
+  }) {
+    const options = {
```

Comment:
Extract this out to a constant `.json` file so it's clear where the configuration is done.

Reply:
Done